fix: add support for CORS headers and pre-flight request

lionello opened the pull request:
This PR implements basic Cross-Origin Resource Sharing (CORS) support to the RPC server, as per the spec at https://www.w3.org/TR/cors/#resource-requests . The spec has been quoted verbatim in the source code for easier validation and maintenance of the code.

* added support for OPTIONS HTTP method
* interpret CORS request headers for pre-flight requests
* set CORS response headers
* two test cases: standard CORS request, and pre-flight request

In practice this PR allows the REST interface to be used directly from a browser. 

All the existing restrictions to the REST interface still apply: IP subnet, port, username, password. For this reason this PR doesn't explicitly check the request's `Origin` with a whitelist.

schancel commented:
What's the desire to use a the RPC interface from a web browser?  This seems like it could open up the same security issue that just it Electrum.  I could see this being useful as a non-default option.

schancel commented on src/httprpc.cpp:
```diff
@@ -152,6 +152,124 @@ static bool RPCAuthorized(const std::string &strAuth,
                             const std::string &) {
+    // First, check and/or set CORS headers according to
+    // https://www.w3.org/TR/cors/#resource-requests
+    do {
```
What is this loop for?

lionello replied:
Using `do { ... } while (false);` to enable the use of `break`.

schancel commented:
@lionello   Hi.  I see you responded, but I'm still curious as to what your use case is?

I'd personally like to see:
- [ ] That loop factored out as another function with `return` instead of `break`
- [ ] Add a flag called `--cors` defaulted to 0

lionello commented:
@schancel Done. I've added the flag `-rpccorsdomain` to match the existing options (and similar option in Ethereum's Geth.)

I've also added tests for that flag.

As for your question: we have an internal website that we use to sign transactions with hardware wallets. We use RPC to send out the signed transactions. We'd prefer to do this without a service (which needs audits, security updates, etc..)

schancel commented:
Hi @lionello , This looks great.  Thank you for doing this.  We don't actually merge code on github, and I haven't made time to move this over to phabricator yet.

I want to speed up the process, I invite you to submit a diff over on https://reviews.bitcoinabc.org 

If you need some help with how, please reach out here and I'll help you out.

Otherwise, your code will get migrated over by me soon. :)

lionello commented:
Ported to https://reviews.bitcoinabc.org/D1112

@schancel Please check whether I did the right thing. Not sure about the repo rABC vs rSTAGING.

schancel commented:
Thank you for getting this moved over to phabricator.

schancel commented:
Hey @lionello are you able to receive emails from phabricator?  I left some comments over there, but overall the commit looks good.   Could you possibly join our telegram group mentioned in the CONTRIBUTING.md guide?  Might make it easier to correspond.

lionello commented:
@schancel I've updated the patch for D1112 and joined the Telegram group.
